### PR TITLE
ANN&GRAM: Add `async/await` and `try` support

### PR DESCRIPTION
**Pull request by mchernyavsky**
Closes https://github.com/intellij-rust/intellij-rust/issues/2400. Relates to https://github.com/intellij-rust/intellij-rust/issues/2622.

### CODE REVIEW

File: src/main/kotlin/org/rust/lang/core/parser/RustParserUtil.kt [hunk 1]
```diff
@@ -295,6 +295,17 @@ object RustParserUtil : GeneratedParserUtilBase() {
     @JvmStatic
     fun dynKeyword(b: PsiBuilder, level: Int): Boolean = contextualKeyword(b, "dyn", DYN)
 
+    @JvmStatic
+    fun asyncKeyword(b: PsiBuilder, level: Int): Boolean = contextualKeyword(b, "async", ASYNC)
+
+    @JvmStatic
+    fun asyncKeywordForBlocks(b: PsiBuilder, level: Int): Boolean =
+        contextualKeyword(b, "async", ASYNC, setOf(LBRACE, MOVE, OR), invertRequiredNextElements = false)
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/lang/core/parser/RustParserUtil.kt]: `tokenSetOf`? And move it to a constant

**Comment by surdus** [comment]
Tried this changes. Found that it shows errors for await macro `await!(...)` and for combination unsafe and async in function declaration `unsafe async fn foo() {`

File: src/test/kotlin/org/rust/ide/annotator/RsEdition2018KeywordsAnnotatorTest.kt [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+    @MockEdition(CargoWorkspace.Edition.EDITION_2018)
+    fun `test async in edition 2018`() = checkErrors("""
+        async fn foo() {}
+
+        fn main() {
+            async { 42 };
+            async || { 42 };
+            async move || { 42 };
+        }
```
Comment (gfreezy) [inline src/test/kotlin/org/rust/ide/annotator/RsEdition2018KeywordsAnnotatorTest.kt]:
```
async move {}
```

should be supported.

https://github.com/rust-lang/rfcs/blob/master/text/2394-async_await.md#initialization-pattern

File: src/main/grammars/RustParser.bnf [hunk 3]
```diff
@@ -905,15 +911,15 @@ Expr ::= RetExpr
...
+  | '-' | '*' | '!' | '&' | move | LitExpr | while | if | for | continue | break  | loop | match | unsafe | yield | "async" | "try"
 
 // https://github.com/rust-lang/rfcs/blob/master/text/0092-struct-grammar.md
 NoStructLitExpr ::= <<structLiterals 'OFF' <<stmtMode 'OFF' Expr>> >> { elementType = Expr }
 AnyExpr ::= <<structLiterals 'ON' <<stmtMode 'OFF' Expr>> >> { elementType = Expr }
 StmtModeExpr ::= <<structLiterals 'ON' <<stmtMode 'ON' Expr>> >> { elementType = Expr }
 StmtModeExprOff ::= <<stmtMode 'OFF' Expr>> { elementType = Expr }
 
-BlockExpr ::= OuterAttr* [ unsafe &'{' ] SimpleBlock {
+BlockExpr ::= OuterAttr* [ (unsafe | asyncBlock | try) &'{' ] SimpleBlock {
```
Comment (gfreezy) [inline src/main/grammars/RustParser.bnf]: maybe `asyncBlock move?` is enough?

**Comment by gilescope** [comment]
🎅 This is on my xmas list :-)

**Comment by mchernyavsky** [comment]
@gfreezy Can you provide some example code? Auto-completion & resolve work for me:

* Edition 2015:
Example

* Edition 2018:
Example

**Comment by imkow** [comment]
Is this in nightly release?

**Comment by Undin** [comment]
@imkow no, because this PR is not merged yet

**Comment by mchernyavsky** [comment]
> @Undin when will this be merged?

@vlad20012 ping

File: src/main/grammars/RustParser.bnf [hunk 4]
```diff
@@ -245,6 +247,9 @@ private default_ ::= <<defaultKeyword>>
 private union ::= <<unionKeyword>>
 private auto ::= <<autoKeyword>>
 private dyn ::= <<dynKeyword>>
+private async ::= <<asyncKeyword>>
+private asyncBlock ::= <<asyncBlockKeyword>>
+private try ::= <<tryKeyword>>
```
Comment (Undin) [inline src/main/grammars/RustParser.bnf]: Let's add parser tests related to grammar changes

File: src/main/kotlin/org/rust/lang/core/parser/RustParserUtil.kt [hunk 5]
```diff
@@ -343,9 +360,17 @@ object RustParserUtil : GeneratedParserUtilBase() {
         tokenType == RsElementTypes.MACRO_EXPR &&
             '}' == b.originalText.subSequence(startOffset, endOffset).findLast { it == '}' || it == ']' || it == ')' }
 
-    private fun contextualKeyword(b: PsiBuilder, keyword: String, elementType: IElementType): Boolean {
+    private fun contextualKeyword(
+        b: PsiBuilder,
+        keyword: String,
+        elementType: IElementType,
+        requiredNextElements: TokenSet = DEFAULT_REQUIRED_NEXT_ELEMENTS,
```
Comment (Undin) [inline src/main/kotlin/org/rust/lang/core/parser/RustParserUtil.kt]: `required` in variable name is quite confusing. Just `nextElements` is better, in my opinion

File: src/main/kotlin/org/rust/lang/core/parser/RustParserUtil.kt [hunk 6]
```diff
@@ -343,9 +360,17 @@ object RustParserUtil : GeneratedParserUtilBase() {
...
+        invertRequiredNextElements: Boolean = true
```
Comment (Undin) [inline src/main/kotlin/org/rust/lang/core/parser/RustParserUtil.kt]: According to the prev comment, we should find a better name for this flag too but currently, I can come up with nothing suitable

**Comment by gfreezy** [comment]
@mchernyavsky 
image

`id` is not identified as a function.

**Comment by mchernyavsky** [comment]
> `id` is not identified as a function.

@gfreezy What do you mean? On attached screenshot everything seems ok.

File: src/main/kotlin/org/rust/lang/core/parser/RustParserUtil.kt [hunk 7]
```diff
@@ -343,9 +360,17 @@ object RustParserUtil : GeneratedParserUtilBase() {
         tokenType == RsElementTypes.MACRO_EXPR &&
             '}' == b.originalText.subSequence(startOffset, endOffset).findLast { it == '}' || it == ']' || it == ')' }
 
-    private fun contextualKeyword(b: PsiBuilder, keyword: String, elementType: IElementType): Boolean {
+    private fun contextualKeyword(
+        b: PsiBuilder,
+        keyword: String,
+        elementType: IElementType,
+        nextElements: TokenSet = DEFAULT_NEXT_ELEMENTS,
+        invertNextElements: Boolean = true
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/lang/core/parser/RustParserUtil.kt]: I think we can make the function signature much more cleaner and generic. Let's just accept a lambda that accepts next token as a parameter and return boolean. So its call site will looks like `contextualKeyword(b, "try", TRY) { it in TRY_NEXT_ELEMENTS }`

File: src/main/kotlin/org/rust/lang/core/psi/ext/RsFunction.kt [hunk 8]
```diff
@@ -94,6 +94,8 @@ abstract class RsFunctionImplMixin : RsStubbedNamedElementImpl<RsFunctionStub>,
     final override val innerAttrList: List<RsInnerAttr>
         get() = block?.innerAttrList.orEmpty()
 
+    override fun getNameIdentifier(): PsiElement? = findLastChildByType(RsElementTypes.IDENTIFIER)
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/lang/core/psi/ext/RsFunction.kt]: Hmm, looks like it isn't needed for async feature. Is it tested?

Comment (mchernyavsky) [reply src/main/kotlin/org/rust/lang/core/psi/ext/RsFunction.kt]: True, this is no longer needed and was left from the previous implementation. Removed.

**Comment by Undin** [comment]
Note, this PR adds only syntax support
Type inference for `try` expression and all async related things is not implemented yet

**Comment by imkow** [comment]
you didn't cover  ` await!(async { /* ... */ }); ` situation.

**Pull request merged**